Sources/Commands/SwiftTestTool.swift

```diff
@@ -353,6 +332,181 @@ public struct SwiftTestTool: SwiftTool {
         // Parse json and return TestSuites.
         return try TestSuite.parse(jsonString: data)
     }
+
+    /// A class to run tests on a XCTest binary.
+    ///
+    /// Note: Executes the XCTest with inherited environment as it is convenient to pass senstive
+    /// information like username, password etc to test cases via enviornment variables.
+    final class TestRunner {
+        /// Path to valid XCTest binary.
+        private let path: AbsolutePath
+
+        /// Arguments to pass to XCTest if any.
+        private let xctestArg: String?
+
+        /// Creates an instance of TestRunner.
+        ///
+        /// - Parameters:
+        ///     - path: Path to valid XCTest binary.
+        ///     - xctestArg: Arguments to pass to XCTest.
+        init(path: AbsolutePath, xctestArg: String? = nil) {
+            self.path = path
+            self.xctestArg = xctestArg
+        }
+
+        /// Constructs arguments to execute XCTest.
+        private func args() -> [String] {
+            var args: [String] = []
+          #if os(macOS)
+            args = ["xcrun", "xctest"]
+            if let xctestArg = xctestArg {
+                args += ["-XCTest", xctestArg]
+            }
+            args += [path.asString]
+          #else
+            args += [path.asString]
+            if let xctestArg = xctestArg {
+                args += [xctestArg]
+            }
+          #endif
+            return args
+        }
+
+        /// Current inherited enviornment variables.
+        private var environment: [String: String] {
+          #if os(Linux)
+            return ProcessInfo.processInfo().environment
+          #else
+            return ProcessInfo.processInfo.environment
+          #endif
+        }
+
+        /// Executes the tests without printing anything on standard streams.
+        ///
+        /// - Returns: A tuple with first bool member indicating if test execution returned code 0
+        ///            and second argument containing the output of the execution.
+        func test() -> (Bool, String) {
+            var output = ""
+            var success = true
+            do {
+                try popen(args(), redirectStandardError: true, environment: environment) { line in
+                    output += line
+                }
+            } catch {
+                success = false
+            }
+            return (success, output)
+        }
+
+        /// Executes and returns execution status. Prints test output on standard streams.
+        func test() -> Bool {
+            let result: Void? = try? system(args(), environment: environment)
+            return result != nil
+        }
+    }
+
+    /// A class to run tests in parallel.
+    final class ParallelTestRunner {
+        /// Path to XCTest binary.
+        private let testPath: AbsolutePath
+
+        /// The producer queue containing list of tests to run.
+        private let producer = SynchronizedQueue<String?>()
```

Comment:
I would give this a name more indicative of its function rather than its role (`pendingTestsQueue` or something). Same for the consumer queue below.